ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java

```diff
@@ -1163,9 +1167,15 @@ private static class MaterializationRebuildLockHeartbeater implements Runnable {
     public void run() {
       LOG.trace("Heartbeating materialization rebuild lock for {} for query: {}",
           AcidUtils.getFullTableName(dbName, tableName), queryId);
-      boolean refreshed;
+      boolean refreshed = false;
       try {
-        refreshed = txnMgr.heartbeatMaterializationRebuildLock(dbName, tableName, txnId);
+        if (!txnMgr.getValidTxns(EXCLUDE_TXN_TYPES).isTxnAborted(txnId)) {
```

Comment:
That change improves nothing speaking of the original idea that I wrote before. And also, it introduce a new problem. 

I wrote this originally: 
> Selecting all the transactions in a heartbeat, just to check if the txn is aborted looks like an unnecessary resource consumption that we could avoid.

If you check the implementation in TxnHandler, you will see that excluding transaction types is just a cosmetic change: you save some bytes in API calls but still, you select all the transactions from the metastore database. So now, you can only save some network bandwith between metastore and HS2. 
You can see the select statements in `GetOpenTxnsListHandler` and the filtering logic in `OpenTxnList.toOpenTxnsResponse`. 

Secondly, that implementation defines everything as an exclusion. And keeps one type. That makes hard to introduce new transaction types. Imagine: you have to introduce a new transaction type, 2 years from now. Will you remember that you have to add that type in an exclusion list at a materialized view lock heartbeater? 
Imagine an other approach: what about using inclusions, not exclusions and saying we only need `TxnType.MATER_VIEW_REBUILD`? 

At this point, I don't know what is the best solution: keeping this logic is a no from my side as it increases the complexity without having any performance boost. And also introduces a possible problem in case we want to extend `TxnType`. 
But I also now that using an inclusion, not an exclusion, leads way more code changes: changes in the sql string , modifying the TxnHandler interface, re-generating Thrift stuff... It is a tedious task to do. But without it, I see more problem with that change than value.

Reply:
Totally agree with you. I will do one thing, I will create a separate issue to handle MV rebuild cleaner task/heart beater issue and as a part of this issue I will fix only current issue(Removing locak entry from MATERIALIZATION_REBUILD_LOCKS table in case transaction is aborted.)

Reply:
Created [HIVE-28563](https://issues.apache.org/jira/browse/HIVE-28563) to fix the cleaner/heartbeater related issue.